src/engine/controls/bpmcontrol.cpp

```diff
@@ -48,149 +42,163 @@ BpmControl::BpmControl(QString group,
+    connect(m_pBeatsTranslateMatchAlignment.get(),
+            &ControlObject::valueChanged,
+            this,
+            &BpmControl::slotBeatsTranslateMatchAlignment,
             Qt::DirectConnection);
 
-    connect(&m_tapFilter, &TapFilter::tapped,
-            this, &BpmControl::slotTapFilter,
+    connect(&m_tapFilter,
+            &TapFilter::tapped,
+            this,
+            &BpmControl::slotTapFilter,
             Qt::DirectConnection);
 
     // Measures distance from last beat in percentage: 0.5 = half-beat away.
-    m_pThisBeatDistance = new ControlProxy(group, "beat_distance", this);
-    m_pSyncMode = new ControlProxy(group, "sync_mode", this);
-}
-
-BpmControl::~BpmControl() {
-    delete m_pLocalBpm;
-    delete m_pEngineBpm;
-    delete m_pButtonTap;
-    delete m_pButtonSync;
-    delete m_pButtonSyncPhase;
-    delete m_pButtonSyncTempo;
-    delete m_pTranslateBeats;
-    delete m_pBeatsTranslateMatchAlignment;
-    delete m_pTranslateBeatsEarlier;
-    delete m_pTranslateBeatsLater;
-    delete m_pAdjustBeatsFaster;
-    delete m_pAdjustBeatsSlower;
+    m_pThisBeatDistance =
+            make_parented<ControlProxy>(group, "beat_distance", this);
+    m_pSyncMode = make_parented<ControlProxy>(group, "sync_mode", this);
 }
 
 double BpmControl::getBpm() const {
     return m_pEngineBpm->get();
 }
 
 void BpmControl::slotAdjustBeatsFaster(double v) {
-    mixxx::BeatsPointer pBeats = m_pBeats;
-    if (v > 0 && pBeats && (pBeats->getCapabilities() & mixxx::Beats::BEATSCAP_SETBPM)) {
-        double bpm = pBeats->getBpm();
-        double adjustedBpm = bpm + kBpmAdjustStep;
-        pBeats->setBpm(adjustedBpm);
+    if (v > 0 && m_pBeats) {
+        auto bpm = m_pBeats->getGlobalBpm();
+        auto adjustedBpm = bpm + kBpmAdjustStep;
+        m_pBeats->setBpm(adjustedBpm);
     }
 }
 
 void BpmControl::slotAdjustBeatsSlower(double v) {
-    mixxx::BeatsPointer pBeats = m_pBeats;
-    if (v > 0 && pBeats && (pBeats->getCapabilities() & mixxx::Beats::BEATSCAP_SETBPM)) {
-        double bpm = pBeats->getBpm();
-        double adjustedBpm = math_max(kBpmAdjustMin, bpm - kBpmAdjustStep);
-        pBeats->setBpm(adjustedBpm);
+    if (v > 0 && m_pBeats) {
+        auto bpm = m_pBeats->getGlobalBpm();
+        auto adjustedBpm = mixxx::Bpm(math_max(kBpmAdjustMin, (bpm - kBpmAdjustStep).getValue()));
+        m_pBeats->setBpm(adjustedBpm);
     }
 }
 
 void BpmControl::slotTranslateBeatsEarlier(double v) {
-    mixxx::BeatsPointer pBeats = m_pBeats;
-    if (v > 0 && pBeats &&
-            (pBeats->getCapabilities() & mixxx::Beats::BEATSCAP_TRANSLATE)) {
-        const int translate_dist = getSampleOfTrack().rate * -.01;
-        pBeats->translate(translate_dist);
+    if (v > 0 && m_pBeats) {
+        const mixxx::FrameDiff_t translateDistFrames = samplesToFrames(
+                getSampleOfTrack().rate * -kSmallBeatsTranslateFactor);
+        m_pBeats->translate(translateDistFrames);
```

Comment:
It was a design choice to make the sample rate unavailable to the public interface of the Beats class, since it is an internal implementation detail for it, and the sample rate is a property of the Track, not Beats.

Reply:
Instead of revealing the sample rate those operations should be added as methods to `Beats` where they have access to the internal sample rate. The translation distance then needs to be specified in time units and is mapped to frames internally.

Reply:
Yes right, since bBeats is aware of the seconds representation it can take kSmallBeatsTranslateFactor directly.